api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,418 @@
+/**
+ *
+ */
+package org.commcare.api.persistence;
+
+import org.javarosa.core.services.PrototypeManager;
+import org.javarosa.core.services.storage.EntityFilter;
+import org.javarosa.core.services.storage.IStorageUtilityIndexed;
+import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.services.storage.StorageFullException;
+import org.javarosa.core.util.InvalidIndexException;
+import org.javarosa.core.util.externalizable.DeserializationException;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Vector;
+
+/**
+ * IStorageIndexedUtility implemented on SQLite using JDBC. Contains all the functionality
+ * for interacting with the user's SQLite representation.
+ *
+ * @author wspride
+ */
+public class SqlIndexedStorageUtility<T extends Persistable> implements IStorageUtilityIndexed<T>, Iterable<T> {
+
+    private final Class<T> prototype;
+    private final String tableName;
+    private final String userName;
+
+    public SqlIndexedStorageUtility(Class<T> prototype, String userName, String tableName) {
+        this.tableName = tableName;
+        this.userName = userName;
+        this.prototype = prototype;
+        tryCreateTable();
+    }
+
+    Connection getConnection() {
+        try {
+            Class.forName("org.sqlite.JDBC");
+            return DriverManager.getConnection("jdbc:sqlite:" + this.userName + ".db");
```

Comment:
I don't know much about this, but have you looked into using `DataSource` instead of `DriverManager`? It supposedly uses a connection pool that is faster if you are opening and closing connections often, which we do here (and is the right way to do DB connections [according to this](http://stackoverflow.com/a/1684020))